Feat(CORE-1255): Add pipelines code exec docs (2) (#1099)
gruntwork-io/docs: docs/pipelines/architecture/index.md

# Architecture

> There are three components in pipelines - the orchestrator, the dispatcher, and the executor. The orchestrator determines the jobs that needs to be executed, the dispatcher takes each job as input and calls the executor then reports the status of the job. Each component is explained in detail in sections below.

Reviewer: I realizes this touches on the executor, but with three components, I think it would read better if all three received a clause in the sentence. For example:

```suggestion
There are three components in pipelines - the orchestrator, the dispatcher, and the executor. The orchestrator determines the jobs that needs to be executed; the dispatcher calls the executor with each job and reports on their status; and the executor actually executes the jobs and makes any necessary updates in AWS.
```

## Orchestrator

> The orchestrator identifies each infrastructure change in the Pull Request or git commit, classifies the type of infrastructure change it is (e.g. AccountAdded, ModuleChanged, EnvCommonChanged) and determines the right pipelines actions (e.g., terragrunt plan, apply, or destroy) to run based on that infrastructure change. For each infrastructure change, it calls the dispatcher.

Reviewer: ```suggestion
The orchestrator identifies each infrastructure change in the pull request or git commit, classifies the type of infrastructure change it is (e.g. AccountAdded, ModuleChanged, EnvCommonChanged), and determines the right pipelines actions (e.g., `terragrunt plan`, `apply`, or `destroy`) to run based on that infrastructure change. For each infrastructure change, it calls the dispatcher.
```

Are we proposing "AccountAdded" etc. as formal terms? I'm wondering about the CamelCase here, or whether we should just use "e.g. account added, module changed, module defaults changed, etc." In either case, we shouldn't use "EnvCommonChanged" anymore.

Author: I like them as formal terms, yes. It introduces the infra change type as a first class concept. I can change the env common bit.

## Dispatcher

> The dispatcher takes an infrastructure change, a pipelines action, and parameters as input. Then, it calls the executor to run the action on the infrastructure change. Once complete, the dispatcher creates a Pull Request comment (for PR workflows) or a GitHub issue (for pushes to main) with a link to the GitHub logs

Reviewer: ```suggestion
The dispatcher takes an infrastructure change, a pipelines action, and parameters as input. Then, it calls the executor to run the action on the infrastructure change. Once complete, the dispatcher creates a pull request comment (for PR workflows) or a GitHub issue (for pushes to main) with a link to the GitHub logs.
```

## Executor

> The executor takes a pipeline action and infra change as input then performs the action on the infra change. In some cases, like responding to AccountAdded events on merge, the executor will create a Pull Request in the infrastructure-live repository with additional IaC code to provision additional resources.

Reviewer: ```suggestion
The executor takes a pipeline action and infra change as input then performs the action on the infra change. In some cases, like responding to AccountAdded events on merge, the executor will create a pull request in the `infrastructure-live` repository with additional IaC code to provision additional resources.
```

## Execution Flow

Pipelines starts with an event in GitHub - a Pull Request being created, synchronized, or re-opened, and a PR being merged to main (e.g., a push to main). From there, the orchestrator determines the infra-change set and appropriate action to run for each infra-change. The dispatcher then sends the information to the dispatcher, which calls the executor and waits for it to complete. Once complete, the status of the execution is reported to the calling Pull Request in a comment or as a GitHub issue if an apply or destroy fails.

> ![Gruntwork Pipelines Execution Flow](/img/pipelines/how-it-works/pipelines_execution_flow.png)

Reviewer: I have a bunch of comments on the diagram but I think it would be easier to provide them in context, rather than in the PR. At a high level, I think this might be slightly too complicated. We want to convey the overall architecture of the 2 repo process, but I think we could simplify this.

Author: I sent you a link to the diagrams in Slack. We can simplify.

Author: Simplified in [bea2047](https://github.com/gruntwork-io/docs/pull/1099/commits/bea2047e0a3f03c389336aa77ce4e572ab1e1f7f) lmk your thoughts
